src/sugar3/graphics/texteditor.py

```diff
@@ -0,0 +1,181 @@
+# Copyright (C) 2015, Batchu Venkat Vishal
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+# Boston, MA 02111-1307, USA.
+
+import logging
+
+from gi.repository import Gtk
+from sugar3.presence.wrapper import CollabWrapper
+
+'''
+The collabtexteditor moudle provides a text editor widget 
+which can be included in any activity and then multiple 
+users can collaborate and edit together in the editor.
+'''
+
+
+class CollabTextEditor(Gtk.Box):
+    '''
+    A CollabTextEditor widget is a adjustable text editor which
+    can be placed on an activity screen.
+
+    The `changed` signal is usually emitted when the text in the
+    editor is changed by a user.
+    The `message` signal is usually emitted when another user makes
+    changes in the text editor, so they are reflected in your editor.
+
+    The widget can be embedded in a window which can be displayed.
+    Example usage:
+        editorinstance = CollabTextEditor(self)
+        scrolled_window.add(editorinstance)
+        scrolled_window.show()
+
+    '''
+
+    def __init__(self, activity, editor_id):
+        Gtk.Box.__init__(self)
+        self._id = editor_id
+        self._callbacks_status = True
+        self._collab = CollabWrapper(activity)
+        self._collab.connect('message', self.__message_cb)
+        self._collab.connect('buddy-joined', self.__buddy_joined_cb)
+        self._collab.setup()
+        self.textview = Gtk.TextView()
+        self.textview.set_editable(True)
+        self.textview.set_cursor_visible(True)
+        self.textview.set_wrap_mode(Gtk.WrapMode.WORD)
+        self.textbuffer = self.textview.get_buffer()
+        self.textbuffer.connect('insert-text', self.__text_buffer_inserted_cb)
+        self.textbuffer.connect('delete-range', self.__text_buffer_deleted_cb)
+        self.textbuffer.set_text("")
+        self.textview.show()
+        self.pack_start(self.textview, expand=True, fill=True, padding=0)
+        self.has_initialized = False
+        self.show()
+
+    '''
+    The leader of the activity sends the latest content to another buddy who joins.
+    '''
+    def get_data(self):
```

Comment:
This get_data and set_data functions must be present in the activity itself and not the widget, but if I put them in the activity they are not connected to collab so errors occur, what should I do?